### ISSUE TITLE
Progress on adding syscall to lind-wasm

### ISSUE THREAD

**Issue by robinyuan1002** [Lind-Project/lind-wasm#12]
Track the progress on adding syscalls to Lind-Wasm

**Comment by robinyuan1002** [comment]
I ran all the test cases of Lind, and basically none of them could compile. Mainly due to the lack of some header files, # include <linux/errno.h>, # include <linux/falloc.h>, # include <asm/socket.h>. I'll try to find and install it.

**Comment by robinyuan1002** [comment]
1. Most of the tests can be compiled, but I encountered a problem in <asm/socket.h>.

wasm-ld: error: /tmp/uds-socketselect-63ca2b.o: undefined symbol: connect
wasm-ld: error: /tmp/uds-socketselect-63ca2b.o: undefined symbol: bind
wasm-ld: error: /tmp/uds-socketselect-63ca2b.o: undefined symbol: accept
wasm-ld: error: /home/lind-wasm/glibc/sysroot/lib/wasm32-wasi/libc.a(nanosleep.o): undefined symbol: __clock_nanosleep_time64
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)


2. I started running these tests. Try modifying dispatch.rs or glibc/sysdeps if I encountered problems.

**Comment by rennergade** [comment]
Where is asm/socket.h included?

**Comment by robinyuan1002** [comment]
uds-nb-select.c, uds-serverclient.c, uds-socketselect.c

**Comment by rennergade** [comment]
I don't see any of those declared in those programs. Are you talking about sys/socket.h?

**Comment by robinyuan1002** [comment]
yes, in sys/socket.h there is a header file called bit/socket.h, and in bit/socket.h there is a header file called asm/socket.h

**Comment by rennergade** [comment]
Okay, so these are included within the libraries so its entirely a glibc issue. Thats good to know.

**Comment by rennergade** [comment]
So for NaCl, asm/socket.h is actually an include file generated by the nacl-toolchain and is platform specific. @yzhang71 are there wasm toolchain generated include files?

**Comment by yzhang71** [comment]
No, we don't have asm/socket.h generated, but we have bit/socket.h generated, and there is condition to use asm/socket.h
```
#ifdef __USE_MISC
# include <bits/types/time_t.h>
# include <asm/socket.h>
#else
# define SO_DEBUG 1
# include <bits/socket-constants.h>
#endif
```

**Comment by qianxichen233** [comment]
For the undefined symbol error, I just found out that it was due to `bind.c` failing to compile and not being successfully turned into `bind.o`. The error message was buried among a sea of warnings and other errors during the make process, so we did not notice it previously. I guess that's the same for connect.c and accept.c. I'll try to fix this later

**Comment by rennergade** [comment]
> No, we don't have asm/socket.h generated, but we have bit/socket.h generated, and there is condition to use asm/socket.h
> 
> ```
> #ifdef __USE_MISC
> # include <bits/types/time_t.h>
> # include <asm/socket.h>
> #else
> # define SO_DEBUG 1
> # include <bits/socket-constants.h>
> #endif
> ```

IIRC bits/socket.h is a file in libc. What I'm saying is that asm/socket.h is not a libc file. It's a platform file.

For example on my system its located at `/usr/x86_64-linux-gnux32/include/asm/socket.h`

**Comment by robinyuan1002** [comment]
Update
There is no problem with open, read and write.

Problem
When I was testing read.c, I encountered a problem. The isprint() function returned false no matter how much input was entered. This seemed to be a problem, but it did not affect the test very much.

**Comment by rennergade** [comment]
@qianxichen233 found that copying the system includes into the wasm folder ie `/usr/x86_64-linux-gnux32/include/asm/socket.h` fixes the builds. We need to:

1. Audit the include folder to make sure its just defined constants and/or code we can compile.
2. Lets ask the people on Zulip if wasm generates its own system includes as above (mention that NaCl does). If they do I'd prefer to use those.

**Comment by rennergade** [comment]
@yzhang71 Can you point me to the wasm generated headers you were talking about in the meeting?

**Comment by yzhang71** [comment]
Hi Nick, here is the header files that we are currently using: https://github.com/Lind-Project/glibc/tree/main/include

**Comment by robinyuan1002** [comment]
Hi Nick, I use glibc/target/include which is same with Dennis's link. I look for the header file I need based on the compile error, for example I created a directory called asm-generic and add a file called errno.base.h(Here is the link: https://github.com/torvalds/linux/blob/master/include/uapi/asm-generic/errno-base.h).

**Comment by rennergade** [comment]
Okay that sounds good. Dennis is going to check with Zulip but we can probably use these if they dont have anything. We probably want to move these out of the glibc folder since theyre not technically libc files.

**Comment by robinyuan1002** [comment]
Update:
There is no problem with rename(6), xstat(9), open(10), close(11), read(12), write(13), fstatfs(19), dup(24), dup2(25), getcwd(47), mkdir(131), rmdir(132), chmod(133), fchmod(134)
